app/callbacks/views/sns.py

```diff
@@ -0,0 +1,416 @@
+import datetime
+from functools import lru_cache
+from itertools import chain
+import json
+import logging
+import re
+import sys
+
+import boto3
+from flask import abort, jsonify, current_app, request
+from lxml.etree import fromstring as etree_fromstring, ParseError
+import requests
+
+import validatesns
+
+from dmutils.email.dm_notify import DMNotifyClient
+from dmutils.email.exceptions import EmailError
+from dmutils.timing import logged_duration_for_external_request as log_external_request, logged_duration
+
+from app.callbacks import callbacks
+from app.clam import get_clamd_socket
+
+
+@callbacks.route('/')
+@callbacks.route('')
+def callbacks_root():
+    return jsonify(status='ok'), 200
+
+
+def _get_request_body_json():
+    if request.content_type.split(";")[0] != "application/json":  # allowing for an encoding following the semicolon
+        abort(400, "Unexpected Content-Type, expecting 'application/json'")
+
+    data = request.get_json()
+
+    if data is None:
+        abort(400, "Invalid JSON; must be a valid JSON object")
+
+    return data
+
+
+def _handle_subscription_confirmation(body_json, supported_topic_name):
+    topic_arn = body_json["TopicArn"]
+    topic_name = topic_arn.split(":")[-1]
+    # check it's for the right topic
+    if topic_name != supported_topic_name:
+        current_app.logger.warning(
+            "Received SubscriptionConfirmation request for unrecognized topic name {topic_name} in {topic_arn}",
+            extra={
+                "topic_name": topic_name,
+                "topic_arn": topic_arn,
+            },
+        )
+        abort(400, f"Unrecognized topic name {topic_name!r}")
+
+    # send a request to provided url, ensuring sufficient logging
+    try:
+        with logged_duration(
+            logger=current_app.logger,
+            message=lambda _: (
+                "Made GET request to {target_url} to confirm subscription to topic {topic_arn}"
+                if sys.exc_info()[0] is None else
+                # need to literally format() exception into message as it's difficult to get it injected into extra
+                "Failed to make GET request to {{target_url}} to confirm subscription to {{topic_arn}}: {!r}".format(
+                    sys.exc_info()[1]
+                )
+            ),
+            log_level=logging.INFO,
+            condition=True,
+        ) as log_context:
+            confirmation_url = body_json["SubscribeURL"]
+            log_context.update({
+                "target_url": confirmation_url,
+                "topic_arn": topic_arn,
+            })
+            confirmation_response = requests.get(confirmation_url)
+            confirmation_response.raise_for_status()
+    except requests.exceptions.RequestException:
+        abort(400, f"SubscriptionConfirmation request to {confirmation_url!r} failed")
+
+    # check we received what we expected from url
+    subscription_arn = confirmation_request_id = None
+    try:
```

Comment:
Took me quite a while to work out what was going on here. Might be worth adding more comments?
Also, is there a scenario where subscription_arn is successfully parsed from the xml, but isn't actually an arn? Like if it was `ARN_DOESNT_EXIST`. Or something? Could you check that the retrieved value matches a regex or something?

Reply:
Or is it possible that there will be multiple namespaces? If so, would `subscription_arn` be a list? What happens then?

Reply:
I think ARNs are kinda supposed to be opaque (even though they're not totally) and so I'm just treating it like a string. If they give us a value that isn't an ARN, ... well, ... they shouldn't. It would be against spec. We can't guard against aws breaking all its own rules. I mean whatever we get there kinda *is* our subscription ARN by definition.

The only thing that's going on with the namespaces is I'm trying to get it to leave me alone about namespaces. I'm not doing the `n:` stuff for fun - because the xml has an `xmlns=` attr at the top level, lxml will want me to be explicit about all the namespaces in the xpath expression. Assigning whatever the toplevel namespace uri is to `n` is just my way of making that nonsense as brief as possible.

Reply:
On the other hand, perhaps I shouldn't be lenient, and simply hard code the expected namespace uri and allow it to fail if they ever change that...? There is a degree to which I'm taking the battery out of the smoke alarm here.

Reply:
I think just stick that comment above about namespaces in.